### ISSUE TITLE
forceUpdate in react not a valid comparison?

### ISSUE THREAD

**Issue by greggman** [krausest/js-framework-benchmark#838]
I noticed in react there's this code

```
swapRows = () => {
    const data = this.state.data;
    if (data.length > 998) {
      let temp = data[1];
      data[1] = data[998];
      data[998] = temp;
    }
    this.forceUpdate();
  }
```

Why is that not the same as react hooks? as in

```
  swapRows = () => {
    const data = this.state.data;
    if (data.length > 998) {
      this.setState({ data: [data[0], data[998], ...data.slice(2, 998), data[1], data[999]] })
    }
  }
```

Same for `update` and `remove`

**Comment by greggman** [comment]
Trying it the results seems to show not using `forceUpdate` is better (except for remove)

<img width="284" alt="Screen Shot 2020-12-23 at 13 28 13" src="https://user-images.githubusercontent.com/234804/102959602-50668500-4523-11eb-8695-4e67be5303da.png">

here's the updated code

```
  update = () => {
    const data = this.state.data;
    const newData = data.slice(0);
    for (let i = 0; i < newData.length; i += 10) {
      const r = newData[i];
      newData[i] = { id: r.id, label: r.label + " !!!" };
    }
    this.setState({data: newData});
  }

  remove = (item) => {
    const data = this.state.data;
    const idx = data.indexOf(item);
    this.setState({ data: [...data.slice(0, idx), ...data.slice(idx + 1)]});
  }

  swapRows = () => {
    const data = this.state.data;
    if (data.length > 998) {
      this.setState({ data: [data[0], data[998], ...data.slice(2, 998), data[1], data[999]] });
    }
  }
```

**Comment by ryansolid** [comment]
That's awesome. I think that implementation has been that way for years.  I'm positive that for the longest `forceUpdate` was faster for partial updates but really great to see that it isn't since it feels hacky. Other implementations like Redux and Hooks removed that hackiness. To be fair there are VDOM libraries that benefit from that sort of approach but so much better if we can get rid of it.

**Comment by krausest** [comment]
Yes, it was changed in Oct 2018. 
Would anyone bother if we just removed the class based implementation and just go with the hooks implementation (now that it's even faster)?

**Comment by ryansolid** [comment]
It's a fair question. I still cite this often enough. Out in the community anti-hooks sentiment likes to point to performance and I've needed to answer that question, even wrote an article on it. So I kind of like leaving it around. But maybe we should just make the hooks the official one.. and call out classes instead.

**Comment by oleggrishechkin** [comment]
![Screenshot_20210319_005738](https://user-images.githubusercontent.com/20518092/111703195-57651100-884e-11eb-9194-17c18c2a1d7e.png)

I created some mutable implementation with subscribe and "forceUpdate" with hooks... and i get best results for update and select.

If you create a component for a row tr children it's improve selecting performance, because when you select a row, a row children shouldn't re-render. For best performance you should create a component for id, label, remove button and also spacer

I don't think it's good practice for react, but official react results are terrible, i think pure react solution should be the best of all react implementations like redux, redux-hooks, mobx and other.

**Comment by mrienstra** [comment]
Should this be closed since #869 merged?

**Comment by oleggrishechkin** [comment]
> Should this be closed since #869 merged?

Yeah, now react with classed uses setState with immutable data instead of forceUpdate with mutations for consistency with react-hooks implementation. And It's even faster than forceUpdate.

**Comment by krausest** [comment]
Absolutely!
